### ISSUE TITLE
False negative when calling functions, which accept contravariant generics

### ISSUE THREAD

**Issue by lorenzo-w** [microsoft/pyright#4433]
**Describe the bug**

When supplying a subclass instance as an argument to a function, which expects a generic and contravariant superclass of the former, pyright should raise an error, but it doesn't. See example below for clarity.

**To Reproduce**

Type-check the example code given below.

**Expected behavior**

Pyright should raise an error

**Screenshots or Code**

```python
from typing import Type, Protocol, TypeVar, Generic

T = TypeVar("T", contravariant=True)

class Callback(Protocol[T]):
  def __init__(self, send_type: Type[T]) -> None:
    ...
  
  def send(self, t: T) -> None:
    ...
    
class DummyCallback(Callback[T]):
  def __init__(self, send_type: Type[T]) -> None:
    self.type = send_type
    
  def send(self, t: T) -> None:
    pass
  
class X:
  pass

class Y(X):
  pass

class Z(Y):
  pass

def test_callback(c: Callback[Y]) -> None:
  pass

test_callback(DummyCallback(X))
test_callback(DummyCallback(Y))
test_callback(DummyCallback(Z)) # <--- Should raise an error
```

**VS Code extension or command-line**

Run in VSCode via Pylance version: v2023.1.10 (pyright bbf0ae78)
Type Checking Mode: basic

Edit: added pyright version

**Comment by lorenzo-w** [comment]
Interestingly, when utilizing a standard `Callable` generic, pyright correctly recognizes the error:

```python
from typing import Callable

class X:
  pass

class Y(X):
  pass

class Z(Y):
  pass

def test_callback(c: Callable[[Y], None]) -> None:
  pass

def dummy_callback_x(x: X):
  pass

def dummy_callback_y(y: Y):
  pass

def dummy_callback_z(z: Z):
  pass

test_callback(dummy_callback_x)
test_callback(dummy_callback_y)
test_callback(dummy_callback_z)
```

**Comment by parched** [comment]
Doesn't the covariance of `Type` mean this is fine. i.e.,
```
test_callback(DummyCallback[Y](Z))
```
is fine, but
```
test_callback(DummyCallback[Z](Z))
```
isn't.

Can you provide more context on how `send_type` is used?

**Comment by erictraut** [comment]
Actually, I think the current behavior is correct. Pyright is using bidirectional type inference, and its constraint solver is able to find a solution that meets all the constraints.

Let's simplify the code somewhat, removing the protocol.

```python
from typing import Generic, TypeVar

T = TypeVar("T", contravariant=True)

class Callback(Generic[T]):
    def __init__(self, send_type: type[T]) -> None:
        self.type = send_type

class X: pass
class Y(X): pass
class Z(Y): pass

x: Callback[Y] = Callback(X)
reveal_type(x) # Callback[X]
z: Callback[Y] = Callback(Z)
reveal_type(z) # Callback[Y]
```

Note that the type of `x` after the assignment is narrowed to `Callback[X]`. The type of `z` after assignment is not narrowed to `Callback[Z]`. Rather, it remains `Callback[Y]` because it cannot be narrowed further due to the contravariant type variable.

Note that if you assign `Callback(Z)` to a temporary local variable, then bidirectional type inference no longer applies, and the resulting type is `Callback[Z]` which isn't type compatible with `Callback[Y]`.

```python
z1 = Callback(Z)
reveal_type(z1) # Callback[Z]
z2: Callback[Y] = z1 # Type error
```

I'll note that mypy agrees with pyright here (although mypy appears to have a bug where it doesn't correctly handle `x: Callback[Y] = Callback(X)` without generating an error).

In your second example, `dummy_callback_z` produces an error because bidirectional type inference in this case doesn't allow for a solution. Mypy also agrees with pyright in this case.
